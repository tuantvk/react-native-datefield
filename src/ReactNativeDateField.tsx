import React, { useRef, useState } from 'react';
import {
  View,
  StyleSheet,
  Keyboard,
  TextInput,
  StyleProp,
  ViewStyle,
  TextStyle,
} from 'react-native';
import {
  int,
  formatYearDigits,
  getOnlyNumber,
  getDateDefault,
  daysInMonth,
} from './utils';

export type DateFieldProps = {
  testID?: string;
  containerStyle?: StyleProp<ViewStyle>;
  styleInput?: StyleProp<TextStyle>;
  labelDate?: string;
  labelMonth?: string;
  labelYear?: string;
  defaultValue?: Date;
  onSubmit?(value: Date): void;
  disabled?: boolean;
};

const defaultProps: DateFieldProps = {
  labelDate: 'Date',
  labelMonth: 'Month',
  labelYear: 'Year',
};

const ReactNativeDateField: React.FunctionComponent<DateFieldProps> = ({
  testID,
  containerStyle,
  styleInput,
  labelDate,
  labelMonth,
  labelYear,
  defaultValue,
  onSubmit,
  disabled,
}) => {
  const refMonth = useRef<TextInput>(null);
  const refYear = useRef<TextInput>(null);
  const [dateValue, setDateValue] = useState(getDateDefault(defaultValue));

  const onChangeDate = (value: string) => {
    let date = getOnlyNumber(int(value) > 31 ? '31' : value);
    setDateValue((prev) => ({ ...prev, date }));
    if (date.length === 2) {
      refMonth.current?.focus();
    }
  };

  const onChangeMonth = (value: string) => {
    let month = getOnlyNumber(int(value) > 12 ? '12' : value);
    setDateValue((prev) => ({
      ...prev,
      month,
      date: daysInMonth(dateValue.date, month),
    }));
    if (month.length === 2) {
      refYear.current?.focus();
    }
  };

  const onChangeYear = (value: string) => {
    let current: number = new Date().getFullYear();
    let year = getOnlyNumber(int(value) > current ? current.toString() : value);
    setDateValue((prev) => ({ ...prev, year }));
    if (year.length === 4) {
      onSubmitEditing(year);
      Keyboard.dismiss();
    }
  };

  const onSubmitEditing = (year: string) => {
    onSubmit &&
      onSubmit(new Date(`${year}-${dateValue.month}-${dateValue.date}`));
  };

  const onBlur = () => {
    let current = { ...dateValue };
    if (current.date === '0') {
      current.date = '01';
    }
    if (current.month === '0') {
      current.month = '01';
    }
    if (daysInMonth(current.date, current.month) !== current.date) {
      current.date = daysInMonth(current.date, current.month);
    }
    if (current.year === '0') {
      current.year = `${formatYearDigits(new Date().getFullYear())}`;
    }
    if (current.year.length > 1 && current.year.length < 4) {
      current.year = `${formatYearDigits(int(current.year))}`;
    }
    setDateValue(current);
    onSubmitEditing(current.year);
  };

  return (
    <View {...{ testID }} style={[styles.container, containerStyle]}>
      <TextInput
        value={dateValue.date}
        maxLength={2}
        blurOnSubmit={false}
        returnKeyType="next"
        keyboardType="numeric"
        editable={!disabled}
        placeholder={labelDate}
        style={[styles.input, styleInput]}
        onChangeText={onChangeDate}
        onBlur={onBlur}
        onSubmitEditing={() => refMonth.current?.focus()}
      />
      <TextInput
        ref={refMonth}
        value={dateValue.month}
        maxLength={2}
        blurOnSubmit={false}
        returnKeyType="next"
        keyboardType="numeric"
        editable={!disabled}
        placeholder={labelMonth}
        style={[styles.input, styleInput]}
        onChangeText={onChangeMonth}
        onBlur={onBlur}
        onSubmitEditing={() => refYear.current?.focus()}
      />
      <TextInput
        ref={refYear}
        value={dateValue.year}
        maxLength={4}
        blurOnSubmit={false}
        returnKeyType="done"
        keyboardType="numeric"
        editable={!disabled}
        placeholder={labelYear}
        style={[styles.input, styleInput]}
        onChangeText={onChangeYear}
        onBlur={onBlur}
        onSubmitEditing={() => Keyboard.dismiss()}
      />
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
  },
  input: {
    textAlign: 'center',
  },
});

ReactNativeDateField.defaultProps = defaultProps;
export default ReactNativeDateField;
