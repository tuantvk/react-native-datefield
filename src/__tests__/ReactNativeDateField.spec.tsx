import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import ReactNativeDateField from '../ReactNativeDateField';

describe('DateField component', () => {
  it('should match snapshot', () => {
    const component = shallow(<ReactNativeDateField />);
    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
  });

  describe('Props', () => {
    it('testID', () => {
      const component = shallow(<ReactNativeDateField testID="date_input" />);
      expect(component.length).toBe(1);
      expect(toJson(component)).toMatchSnapshot();
    });

    it('containerStyle', () => {
      const component = shallow(
        <ReactNativeDateField containerStyle={{ marginVertical: 20 }} />
      );
      expect(component.length).toBe(1);
      expect(toJson(component)).toMatchSnapshot();
    });

    it('styleInput', () => {
      const component = shallow(
        <ReactNativeDateField styleInput={{ fontSize: 15 }} />
      );
      expect(component.length).toBe(1);
      expect(toJson(component)).toMatchSnapshot();
    });

    it('labelDate', () => {
      const component = shallow(
        <ReactNativeDateField labelDate="Input date" />
      );
      expect(component.length).toBe(1);
      expect(toJson(component)).toMatchSnapshot();
    });

    it('labelMonth', () => {
      const component = shallow(
        <ReactNativeDateField labelMonth="Input month" />
      );
      expect(component.length).toBe(1);
      expect(toJson(component)).toMatchSnapshot();
    });

    it('labelYear', () => {
      const component = shallow(
        <ReactNativeDateField labelYear="Input year" />
      );
      expect(component.length).toBe(1);
      expect(toJson(component)).toMatchSnapshot();
    });

    it('defaultValue', () => {
      const component = shallow(
        <ReactNativeDateField defaultValue={new Date('2021-03-25')} />
      );
      expect(component.length).toBe(1);
      expect(toJson(component)).toMatchSnapshot();
    });

    it('onSubmit', () => {
      const component = shallow(<ReactNativeDateField onSubmit={() => { }} />);
      expect(component.length).toBe(1);
      expect(toJson(component)).toMatchSnapshot();
    });

    it('disabled', () => {
      const component = shallow(<ReactNativeDateField disabled />);
      expect(component.length).toBe(1);
      expect(toJson(component)).toMatchSnapshot();
    });
  });
});
