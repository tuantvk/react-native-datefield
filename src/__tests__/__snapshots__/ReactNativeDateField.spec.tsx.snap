// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateField component Props containerStyle 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      },
      Object {
        "marginVertical": 20,
      },
    ]
  }
>
  <Component
    allowFontScaling={true}
    blurOnSubmit={false}
    editable={true}
    keyboardType="numeric"
    maxLength={2}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Date"
    rejectResponderTermination={true}
    returnKeyType="next"
    style={
      Array [
        Object {
          "textAlign": "center",
        },
        undefined,
      ]
    }
    underlineColorAndroid="transparent"
    value=""
  />
  <Component
    allowFontScaling={true}
    blurOnSubmit={false}
    editable={true}
    keyboardType="numeric"
    maxLength={2}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Month"
    rejectResponderTermination={true}
    returnKeyType="next"
    style={
      Array [
        Object {
          "textAlign": "center",
        },
        undefined,
      ]
    }
    underlineColorAndroid="transparent"
    value=""
  />
  <Component
    allowFontScaling={true}
    blurOnSubmit={false}
    editable={true}
    keyboardType="numeric"
    maxLength={4}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Year"
    rejectResponderTermination={true}
    returnKeyType="done"
    style={
      Array [
        Object {
          "textAlign": "center",
        },
        undefined,
      ]
    }
    underlineColorAndroid="transparent"
    value=""
  />
</View>
`;

exports[`DateField component Props defaultValue 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      },
      undefined,
    ]
  }
>
  <Component
    allowFontScaling={true}
    blurOnSubmit={false}
    editable={true}
    keyboardType="numeric"
    maxLength={2}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Date"
    rejectResponderTermination={true}
    returnKeyType="next"
    style={
      Array [
        Object {
          "textAlign": "center",
        },
        undefined,
      ]
    }
    underlineColorAndroid="transparent"
    value="25"
  />
  <Component
    allowFontScaling={true}
    blurOnSubmit={false}
    editable={true}
    keyboardType="numeric"
    maxLength={2}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Month"
    rejectResponderTermination={true}
    returnKeyType="next"
    style={
      Array [
        Object {
          "textAlign": "center",
        },
        undefined,
      ]
    }
    underlineColorAndroid="transparent"
    value="3"
  />
  <Component
    allowFontScaling={true}
    blurOnSubmit={false}
    editable={true}
    keyboardType="numeric"
    maxLength={4}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Year"
    rejectResponderTermination={true}
    returnKeyType="done"
    style={
      Array [
        Object {
          "textAlign": "center",
        },
        undefined,
      ]
    }
    underlineColorAndroid="transparent"
    value="2021"
  />
</View>
`;

exports[`DateField component Props disabled 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      },
      undefined,
    ]
  }
>
  <Component
    allowFontScaling={true}
    blurOnSubmit={false}
    editable={false}
    keyboardType="numeric"
    maxLength={2}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Date"
    rejectResponderTermination={true}
    returnKeyType="next"
    style={
      Array [
        Object {
          "textAlign": "center",
        },
        undefined,
      ]
    }
    underlineColorAndroid="transparent"
    value=""
  />
  <Component
    allowFontScaling={true}
    blurOnSubmit={false}
    editable={false}
    keyboardType="numeric"
    maxLength={2}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Month"
    rejectResponderTermination={true}
    returnKeyType="next"
    style={
      Array [
        Object {
          "textAlign": "center",
        },
        undefined,
      ]
    }
    underlineColorAndroid="transparent"
    value=""
  />
  <Component
    allowFontScaling={true}
    blurOnSubmit={false}
    editable={false}
    keyboardType="numeric"
    maxLength={4}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Year"
    rejectResponderTermination={true}
    returnKeyType="done"
    style={
      Array [
        Object {
          "textAlign": "center",
        },
        undefined,
      ]
    }
    underlineColorAndroid="transparent"
    value=""
  />
</View>
`;

exports[`DateField component Props labelDate 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      },
      undefined,
    ]
  }
>
  <Component
    allowFontScaling={true}
    blurOnSubmit={false}
    editable={true}
    keyboardType="numeric"
    maxLength={2}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Input date"
    rejectResponderTermination={true}
    returnKeyType="next"
    style={
      Array [
        Object {
          "textAlign": "center",
        },
        undefined,
      ]
    }
    underlineColorAndroid="transparent"
    value=""
  />
  <Component
    allowFontScaling={true}
    blurOnSubmit={false}
    editable={true}
    keyboardType="numeric"
    maxLength={2}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Month"
    rejectResponderTermination={true}
    returnKeyType="next"
    style={
      Array [
        Object {
          "textAlign": "center",
        },
        undefined,
      ]
    }
    underlineColorAndroid="transparent"
    value=""
  />
  <Component
    allowFontScaling={true}
    blurOnSubmit={false}
    editable={true}
    keyboardType="numeric"
    maxLength={4}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Year"
    rejectResponderTermination={true}
    returnKeyType="done"
    style={
      Array [
        Object {
          "textAlign": "center",
        },
        undefined,
      ]
    }
    underlineColorAndroid="transparent"
    value=""
  />
</View>
`;

exports[`DateField component Props labelMonth 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      },
      undefined,
    ]
  }
>
  <Component
    allowFontScaling={true}
    blurOnSubmit={false}
    editable={true}
    keyboardType="numeric"
    maxLength={2}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Date"
    rejectResponderTermination={true}
    returnKeyType="next"
    style={
      Array [
        Object {
          "textAlign": "center",
        },
        undefined,
      ]
    }
    underlineColorAndroid="transparent"
    value=""
  />
  <Component
    allowFontScaling={true}
    blurOnSubmit={false}
    editable={true}
    keyboardType="numeric"
    maxLength={2}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Input month"
    rejectResponderTermination={true}
    returnKeyType="next"
    style={
      Array [
        Object {
          "textAlign": "center",
        },
        undefined,
      ]
    }
    underlineColorAndroid="transparent"
    value=""
  />
  <Component
    allowFontScaling={true}
    blurOnSubmit={false}
    editable={true}
    keyboardType="numeric"
    maxLength={4}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Year"
    rejectResponderTermination={true}
    returnKeyType="done"
    style={
      Array [
        Object {
          "textAlign": "center",
        },
        undefined,
      ]
    }
    underlineColorAndroid="transparent"
    value=""
  />
</View>
`;

exports[`DateField component Props labelYear 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      },
      undefined,
    ]
  }
>
  <Component
    allowFontScaling={true}
    blurOnSubmit={false}
    editable={true}
    keyboardType="numeric"
    maxLength={2}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Date"
    rejectResponderTermination={true}
    returnKeyType="next"
    style={
      Array [
        Object {
          "textAlign": "center",
        },
        undefined,
      ]
    }
    underlineColorAndroid="transparent"
    value=""
  />
  <Component
    allowFontScaling={true}
    blurOnSubmit={false}
    editable={true}
    keyboardType="numeric"
    maxLength={2}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Month"
    rejectResponderTermination={true}
    returnKeyType="next"
    style={
      Array [
        Object {
          "textAlign": "center",
        },
        undefined,
      ]
    }
    underlineColorAndroid="transparent"
    value=""
  />
  <Component
    allowFontScaling={true}
    blurOnSubmit={false}
    editable={true}
    keyboardType="numeric"
    maxLength={4}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Input year"
    rejectResponderTermination={true}
    returnKeyType="done"
    style={
      Array [
        Object {
          "textAlign": "center",
        },
        undefined,
      ]
    }
    underlineColorAndroid="transparent"
    value=""
  />
</View>
`;

exports[`DateField component Props onSubmit 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      },
      undefined,
    ]
  }
>
  <Component
    allowFontScaling={true}
    blurOnSubmit={false}
    editable={true}
    keyboardType="numeric"
    maxLength={2}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Date"
    rejectResponderTermination={true}
    returnKeyType="next"
    style={
      Array [
        Object {
          "textAlign": "center",
        },
        undefined,
      ]
    }
    underlineColorAndroid="transparent"
    value=""
  />
  <Component
    allowFontScaling={true}
    blurOnSubmit={false}
    editable={true}
    keyboardType="numeric"
    maxLength={2}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Month"
    rejectResponderTermination={true}
    returnKeyType="next"
    style={
      Array [
        Object {
          "textAlign": "center",
        },
        undefined,
      ]
    }
    underlineColorAndroid="transparent"
    value=""
  />
  <Component
    allowFontScaling={true}
    blurOnSubmit={false}
    editable={true}
    keyboardType="numeric"
    maxLength={4}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Year"
    rejectResponderTermination={true}
    returnKeyType="done"
    style={
      Array [
        Object {
          "textAlign": "center",
        },
        undefined,
      ]
    }
    underlineColorAndroid="transparent"
    value=""
  />
</View>
`;

exports[`DateField component Props styleInput 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      },
      undefined,
    ]
  }
>
  <Component
    allowFontScaling={true}
    blurOnSubmit={false}
    editable={true}
    keyboardType="numeric"
    maxLength={2}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Date"
    rejectResponderTermination={true}
    returnKeyType="next"
    style={
      Array [
        Object {
          "textAlign": "center",
        },
        Object {
          "fontSize": 15,
        },
      ]
    }
    underlineColorAndroid="transparent"
    value=""
  />
  <Component
    allowFontScaling={true}
    blurOnSubmit={false}
    editable={true}
    keyboardType="numeric"
    maxLength={2}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Month"
    rejectResponderTermination={true}
    returnKeyType="next"
    style={
      Array [
        Object {
          "textAlign": "center",
        },
        Object {
          "fontSize": 15,
        },
      ]
    }
    underlineColorAndroid="transparent"
    value=""
  />
  <Component
    allowFontScaling={true}
    blurOnSubmit={false}
    editable={true}
    keyboardType="numeric"
    maxLength={4}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Year"
    rejectResponderTermination={true}
    returnKeyType="done"
    style={
      Array [
        Object {
          "textAlign": "center",
        },
        Object {
          "fontSize": 15,
        },
      ]
    }
    underlineColorAndroid="transparent"
    value=""
  />
</View>
`;

exports[`DateField component Props testID 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      },
      undefined,
    ]
  }
  testID="date_input"
>
  <Component
    allowFontScaling={true}
    blurOnSubmit={false}
    editable={true}
    keyboardType="numeric"
    maxLength={2}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Date"
    rejectResponderTermination={true}
    returnKeyType="next"
    style={
      Array [
        Object {
          "textAlign": "center",
        },
        undefined,
      ]
    }
    underlineColorAndroid="transparent"
    value=""
  />
  <Component
    allowFontScaling={true}
    blurOnSubmit={false}
    editable={true}
    keyboardType="numeric"
    maxLength={2}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Month"
    rejectResponderTermination={true}
    returnKeyType="next"
    style={
      Array [
        Object {
          "textAlign": "center",
        },
        undefined,
      ]
    }
    underlineColorAndroid="transparent"
    value=""
  />
  <Component
    allowFontScaling={true}
    blurOnSubmit={false}
    editable={true}
    keyboardType="numeric"
    maxLength={4}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Year"
    rejectResponderTermination={true}
    returnKeyType="done"
    style={
      Array [
        Object {
          "textAlign": "center",
        },
        undefined,
      ]
    }
    underlineColorAndroid="transparent"
    value=""
  />
</View>
`;

exports[`DateField component should match snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      },
      undefined,
    ]
  }
>
  <Component
    allowFontScaling={true}
    blurOnSubmit={false}
    editable={true}
    keyboardType="numeric"
    maxLength={2}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Date"
    rejectResponderTermination={true}
    returnKeyType="next"
    style={
      Array [
        Object {
          "textAlign": "center",
        },
        undefined,
      ]
    }
    underlineColorAndroid="transparent"
    value=""
  />
  <Component
    allowFontScaling={true}
    blurOnSubmit={false}
    editable={true}
    keyboardType="numeric"
    maxLength={2}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Month"
    rejectResponderTermination={true}
    returnKeyType="next"
    style={
      Array [
        Object {
          "textAlign": "center",
        },
        undefined,
      ]
    }
    underlineColorAndroid="transparent"
    value=""
  />
  <Component
    allowFontScaling={true}
    blurOnSubmit={false}
    editable={true}
    keyboardType="numeric"
    maxLength={4}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Year"
    rejectResponderTermination={true}
    returnKeyType="done"
    style={
      Array [
        Object {
          "textAlign": "center",
        },
        undefined,
      ]
    }
    underlineColorAndroid="transparent"
    value=""
  />
</View>
`;
